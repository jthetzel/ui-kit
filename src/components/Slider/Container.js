import styled from 'react-emotion';

export default styled.div`
  position: relative;
  text-align: center;
  margin: auto auto ${props => (props.tooltip ? '20px' : 0)};
`;
