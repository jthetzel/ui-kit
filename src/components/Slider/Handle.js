import styled from 'react-emotion';

export default styled.div`
  position: absolute;
  ${({ tooltipPosition }) => (tooltipPosition === 'bottom' ? 'bottom: -15px' : 'top: 0px')};
  left: ${({ position }) => position}%;

  text-align: center;

  transform: translateX(${({ position }) => -position}%);
`;
