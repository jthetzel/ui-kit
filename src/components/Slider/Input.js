import styled from 'react-emotion';

export default styled.input`
  width: 100%;
  height: 34px;
  margin: 0;
   -webkit-appearance: none;
   
  &::-moz-focus-outer {
    border: 0;
  }

  &::-webkit-slider-runnable-track {
      height: 2px;
      background: #000000;
      border: none;
      border-radius: 1px;
  }
  
  &::-webkit-slider-thumb {
      -webkit-appearance: none;
      border: none;
      height: 28px;
      width: 28px;
      background-color: #ffffff;
      border: solid 2px #000000;
      border-radius: 50%;
      margin-top: -14px;
      &:hover {
        cursor: pointer;
      }
  }

  // Firefox
  &:focus {
      outline: none;
  }
  
  &::-moz-range-track {
    height: 2px;
    background: #000000;
    border: none;
    border-radius: 1px;
  }

  &::-moz-range-thumb {
      -webkit-appearance: none;
      border: none;
      height: 28px;
      width: 28px;
      background-color: #ffffff;
      border: solid 2px #000000;
      border-radius: 50%;
      margin-top: -14px;
      &:hover {
          cursor: pointer;
        }
  }

  // IE
  /*hide the outline behind the border*/
  &:-moz-focusring{
      outline: 1px solid white;
      outline-offset: -1px;
  }
  
  &::-ms-track {
      height: 2px;
      background: #000000;
      border: none;
      border-radius: 1px;
  }
  
  &::-ms-fill-lower {
      background: #777;
      border-radius: 10px;
  }
  
  &::-ms-fill-upper {
      background: #ddd;
      border-radius: 10px;
  }
  
  &::-ms-thumb {
      height: 28px;
      width: 28px;
      background-color: #ffffff;
      border: solid 2px #000000;
      border-radius: 50%;
      margin-top: -14px;
      &:hover {
          cursor: pointer;
        }
  }
  
  &:focus::-ms-fill-lower {
      background: #888;
  }
  
  &:focus::-ms-fill-upper {
      background: #ccc;
  }
`;
