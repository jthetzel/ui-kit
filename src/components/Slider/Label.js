import styled from 'react-emotion';

export default styled.div`
  display: block;
  border: 0;
  width: 100%;
  height: 100%;
  text-align: center;
  outline: none;
  cursor: pointer;
  
  &:first-child {
    position: relative;
    text-align: left;
    
    span {
      position: absolute;
      transform: translateX(-50%);
    }
  }
  
  &:last-child {
    position: relative;
    text-align: right;
    
    span {
      position: absolute;
      transform: translateX(-50%);
    }
  }
`;
