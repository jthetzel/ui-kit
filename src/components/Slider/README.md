Button

```js
initialState = { index: 0 };

const options = [
  { value: 'one', label: 'One' },
  { value: 'two', label: 'Two' },
  { value: 'three', label: 'Three' }
];

const onChange = (value) => {
  const index = options.indexOf(value);
  setState({ index })
};

<div>
  Current value: {options[state.index].value}
  <Slider
    onChange={onChange}
    options={options}
    value={options[state.index]}
  />
</div>
```