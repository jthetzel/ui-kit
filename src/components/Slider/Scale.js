import styled from 'react-emotion';

export default styled.div`
  display: flex;
  font-size: 14px;
  font-weight: 500;
  justify-content: space-between;
  padding: 0 12px 0 15px;
`;
