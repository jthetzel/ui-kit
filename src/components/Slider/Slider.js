import React, { PureComponent } from 'react';
import PropTypes from 'prop-types';
import findIndex from '../../utils/findIndex';
import InputRange from '../InputRange';

// Style
import Container from './Container';
import Scale from './Scale';
import Label from './Label';
import Handle from './Handle';

/**
 * @component
 */
class Slider extends PureComponent {
  constructor(props) {
    super(props);

    this.state = {
      index: 0
    };

    this.onChange = this.onChange.bind(this);
    this.onChangeEnd = this.onChangeEnd.bind(this);
    this.onScaleClick = this.onScaleClick.bind(this);
    this._renderOption = this._renderOption.bind(this);
  }

  componentWillMount() {
    const { options, value } = this.props;

    if (!options.length) {
      throw new Error('options is empty');
    }

    const index = findIndex(options, o => o.value === value);

    this.setState({ index });
  }

  onChange(event) {
    this.setState({ index: event.target.value });

    const index = Math.round(event.target.value);
    const { value } = this.props.options[index];

    if (value !== this.props.value) {
      this.props.onChange(value);
    }
  }

  onChangeEnd() {
    const index = Math.round(this.state.index);
    this.setState({ index });
  }

  onScaleClick(index) {
    this.setState({ index });
    const { value } = this.props.options[index];
    this.props.onChange(value);
  }

  _renderOption(option, index) {
    const { ticks } = this.props;
    const isVisible = !ticks || !!(index % ticks);

    return (
      <Label key={option.value} role="button" onClick={this.onScaleClick.bind(null, index)}>
        {isVisible && <span>{option.label}</span>}
      </Label>
    );
  }

  render() {
    const {
      tooltipPosition,
      tooltip,
      options,
      ticks
    } = this.props;

    const { index } = this.state;
    const visibleScale = !ticks || ticks.length !== 0;
    const position = index ? (index / (options.length - 1)) * 100 : 0;
    const roundIndex = Math.round(index);
    const label = options[roundIndex] && options[roundIndex].label;

    return (
      <Container tooltip={tooltip}>
        {
          visibleScale &&
            <Scale>{options.map(this._renderOption)}</Scale>
        }
        <InputRange
          value={this.state.index}
          onChange={this.onChange}
          onClick={this.onChangeEnd}
          onMouseLeave={this.onChangeEnd}
          onTouchEnd={this.onChangeEnd}
          max={options.length - 1}
          step={0.01}
        />
        {
          tooltip &&
            <Handle
              position={position}
              tooltipPosition={tooltipPosition}
            >
              {label}
            </Handle>}
      </Container>
    );
  }
}

Slider.propTypes = {
  options: PropTypes.arrayOf(PropTypes.shape({
    value: PropTypes.any.isRequired,
    label: PropTypes.node
  })).isRequired,
  value: PropTypes.node,
  onChange: PropTypes.func.isRequired,
  tooltip: PropTypes.bool,
  tooltipPosition: PropTypes.oneOf(['bottom', 'top']),
  ticks: PropTypes.number
};

Slider.defaultProps = {
  value: PropTypes.null,
  tooltip: true,
  tooltipPosition: 'bottom',
  ticks: null
};

export default Slider;
