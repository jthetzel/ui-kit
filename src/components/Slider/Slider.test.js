import React from 'react';
import { shallow } from 'enzyme';
import toJson from 'enzyme-to-json';
import Slider from '../Slider';
import Input from '../Slider/Input';

const options = [
  { value: 'one', label: 'One' },
  { value: 'two', label: 'Two' },
  { value: 'three', label: 'Three' }
];

const onChange = value => value;

test('Slider does not crash', () => {
  const wrapper = shallow(<Slider onChange={onChange} options={options} />);
  expect(toJson(wrapper)).toMatchSnapshot();
});

test('Slider with an empty array', () => {
  expect(() => shallow(<Slider onChange={onChange} options={[]} />))
    .toThrowError('options is empty');
});

test('Slider simulate onChange', () => {
  const wrapper = shallow(<Slider
    onChange={onChange}
    options={options}
    value={options[0].value}
  />);
  const rangeInput = wrapper.find(Input).first();
  expect(wrapper.state('index')).toEqual(0);
  rangeInput.simulate('change', { target: { value: 1 } });
  expect(wrapper.state('index')).toEqual(1);
});

test('Slider simulate onChangeEnd', () => {
  const wrapper = shallow(<Slider
    onChange={onChange}
    options={options}
    value={options[0].value}
  />);
  const rangeInput = wrapper.find(Input).first();
  expect(wrapper.state('index')).toEqual(0);
  rangeInput.simulate('change', { target: { value: 0.8 } });
  rangeInput.simulate('click', { target: { value: 0.8 } });
  expect(wrapper.state('index')).toEqual(1);
});

test('Slider simulate onClickScale', () => {
  const wrapper = shallow(<Slider
    onChange={onChange}
    options={options}
    value={options[0].value}
  />);
  const scaleButton = wrapper.find('[role="button"]').last();
  expect(wrapper.state('index')).toEqual(0);
  scaleButton.simulate('click');
  expect(wrapper.state('index')).toEqual(2);
});