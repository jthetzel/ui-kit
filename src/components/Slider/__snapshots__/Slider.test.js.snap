// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Slider does not crash 1`] = `
<Styled(div)>
  <Styled(input)
    max={2}
    onChange={[Function]}
    onClick={[Function]}
    onMouseLeave={[Function]}
    onTouchEnd={[Function]}
    step={0.01}
    type="range"
    value={-1}
  />
  <Styled(div)
    length={3}
  >
    <Styled(div)
      key="one"
      onClick={[Function]}
      role="button"
    >
      <span>
        One
      </span>
    </Styled(div)>
    <Styled(div)
      key="two"
      onClick={[Function]}
      role="button"
    >
      <span>
        Two
      </span>
    </Styled(div)>
    <Styled(div)
      key="three"
      onClick={[Function]}
      role="button"
    >
      <span>
        Three
      </span>
    </Styled(div)>
  </Styled(div)>
</Styled(div)>
`;
